amqprs/src/api/connection.rs

```diff
@@ -1263,6 +1284,24 @@ mod tests {
         }
     }
 
+    #[tokio::test]
+    async fn test_connection_clone_and_drop() {
+        setup_logging();
+        {
+            // test close on drop
+            let args = OpenConnectionArguments::new("localhost", 5672, "user", "bitnami");
+
+            let conn1 = Connection::open(&args).await.unwrap();
+            let conn2 = conn1.clone();
+            tokio::spawn(async move {
+                assert_eq!(true, conn2.is_open());
+            });
+            assert_eq!(true, conn1.is_open());
+        }
+        // wait for finished, otherwise runtime exit before all tasks are done
+        time::sleep(time::Duration::from_millis(100)).await;
```

Comment:
I think the test isn't doing what it should, it should look more like the following:
```suggestion
            let conn1 = Connection::open(&args).await.unwrap();
            let conn2 = conn1.clone();
            {
                let conn3 = conn1;
            }
            assert_eq!(true, conn2.is_open());
        }
```
Reading this PR, I'm not sure that when you clone the `Connection` and then drop a clone, the other instances of the cloned `Connection` will remain open

Reply:
It does test what it should, it tests no matter which one drop first, only last drop will trigger close of connection. So if conn1 drop first, we should see conn2 still open, if conn2 drop first, we can see conn1 still open.

> Reading this PR, I'm not sure that when you clone the `Connection` and then drop a clone, the other instances of the cloned `Connection` will remain ope

You can try to test yourself :)

Reply:
Ah yes indeed I read too fast, you did create a local scope so any of the connections will get dropped before reaching the `time::sleep`.
Also didn't catch that the `DropGuard` was wrapped in an `Arc`, my bad should've read more thoroughly!

Then instead of what I mentioned, which also works, how about:
```rs
        {
            // test close on drop
            let args = OpenConnectionArguments::new("localhost", 5672, "user", "bitnami");

            let conn1 = Connection::open(&args).await.unwrap();
            let conn2 = conn1.clone();
            let t1 = tokio::spawn(async move {
                assert_eq!(true, conn1.is_open());
            });
            let t2 = tokio::spawn(async move {
                assert_eq!(true, conn2.is_open());
            });
            tokio::join!(t1, t2);
        }
```
So that you don't have a `tokio::sleep` in the test?

Reply:
this is good proposal! 
I think it would also get better to call `open_channel` on both conn1 and conn2.
Would you like to submit a new PR to improve this?